spin_coater.c

```diff
@@ -0,0 +1,205 @@
+        reset_dshot_buffers(&ctx->spin_coater);
+#endif
+      }
+    next_delay = make_timeout_time_ms(ctx->spin_coater.rpm_slowdown_update_delay);
+  } else {
+   next_delay = make_timeout_time_ms(1000); 
+  }
+  set_dshot_safe(ctx, ctx->spin_coater.dshot_throttle_val);
+  return next_delay;
+}
+#elif CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_PWM
+static absolute_time_t
+do_pwm_smooth_transition(furnace_context_t* ctx)
+{
+  int direction = ctx->spin_coater.current_rpm < ctx->spin_coater.set_rpm ? 1 : -1;
+
+  ctx->spin_coater.pwm_duty += direction;
+  set_spin_coater_pwm_safe(ctx, ctx->spin_coater.pwm_duty);
+
+  return make_timeout_time_ms(100);
+}
+#endif
+
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+bool repeating_dshot_collect_data_callback(__unused struct repeating_timer *t) {
+    dshot_recv_telemetry((spin_coater_context_t*)t->user_data);
+    return true;
+}
+#endif
+
+static void
+do_spin_coater_throttle_value_update(furnace_context_t *ctx, bool deadline_met)
+{
+    if (deadline_met) {
+        if ((ctx->spin_coater.spin_state == SPIN_STARTED_WITH_TIMER) ||
+            (ctx->spin_coater.spin_state == SPIN_SMOOTH_STOP_REQUESTED)) {
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_PWM
+          ctx->spin_coater_throttle_value_update_deadline = do_pwm_smooth_transition(ctx);
+#elif CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT || CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+          ctx->spin_coater_throttle_value_update_deadline = do_dshot_smooth_transition(ctx);
+#endif
+        } else if(ctx->spin_coater.spin_state == SPIN_STARTED_WITH_FORCE_VALUE) {
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+          ctx->spin_coater.current_rpm = calc_avg_from_raw_dshot_buffer(&ctx->spin_coater);
+#else
+          ctx->spin_coater.current_rpm = rpm_counter_based_on_photo_sensor;
+#endif
+          ctx->spin_coater_throttle_value_update_deadline = make_timeout_time_ms(200);
+        } else {
+          ctx->spin_coater_throttle_value_update_deadline = make_timeout_time_ms(200);
+        }
+    }
+}
+
+void
+init_spin_coater(furnace_context_t *ctx)
+{
+#if CONFIG_SPIN_COATER != CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+  gpio_set_irq_enabled_with_callback(
+    CONFIG_SPIN_COATER_RPM_FEEDBACK_PIN,
+    GPIO_IRQ_EDGE_FALL,
+    true,
+    &rpm_photo_sensor_feedback_irq_handler);
+#endif
+
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_PWM
+  // Tell GPIO 0 and 1 they are allocated to the PWM
+  gpio_set_function(CONFIG_SPIN_COATER_ENGINE_CONTROL_PIN, GPIO_FUNC_PWM);
+  // Find out which PWM slice is connected to GPIO 0 (it's slice 0)
+  uint slice_num = pwm_gpio_to_slice_num(CONFIG_SPIN_COATER_ENGINE_CONTROL_PIN);
+  // Set channel A output high for one cycle before dropping
+  pwm_set_freq_duty(slice_num, PWM_CHAN_A, SPIN_COATER_PWM_490_FREQ, PWM_IDLE_DUTY);
+  // Set initial B output high for three cycles before dropping
+  // Set the PWM running
+  pwm_set_enabled(slice_num, true);
```

Comment:
We already use names pwm_* for heating/furnace PWM. Can we rename to spinpwm_* or something like that?